### PR TITLE
[Security] Removed anonymous in the new security system

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | tbd

This was one of the "Future considerations" of #33558:

> Drop the AnonymousToken and AnonymousAuthenticator: Anonymous authentication has never made much sense and complicates things (e.g. the user can be a string). For access control, an anonymous user has the same meaning as an un-authenticated one (null). This require changes in the AccessListener and AuthorizationChecker and probably also a new Security attribute (to replace IS_AUTHENTICATED_ANONYMOUSLY). Related issues: #34909, #30609

This new experimental system is probably a once-in-a-lifetime change to make this change. @weaverryan and I have had some brainstorming about this. Some reasons why we think it makes 100% sense to do this change:

* From a Security perspective, **a user that is not authenticated is similar to an "unknown" user**: They both have no rights at all.
* **The higher level consequences of the AnonymousToken are confusing and inconsistent**:
  * It's hard to explain people new to Symfony Security that not being logged in still means you're authenticated within the Symfony app
  * To counter this, some higher level APIs explicitly mark anonymous tokens as not being authenticated, see e.g. the [`is_authenticated()` expression language function](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/ExpressionLanguageProvider.php#L33-L37)
  * The anonymous authentication resulted in the `IS_AUTHENTICATED` security attribute being removed from #35854, as there was no clear consensus on what its meaning should be
* **Spring Security, which is where this originated from, makes Anonymous a very special case**:

  > Finally, there is an AnonymousAuthenticationFilter, which is chained after the normal authentication mechanisms and automatically adds an AnonymousAuthenticationToken to the SecurityContextHolder if there is no existing Authentication held there.
  >
  > Note that there is no real conceptual difference between a user who is “anonymously authenticated” and an unauthenticated user. Spring Security's anonymous authentication just gives you a more convenient way to configure your access-control attributes. Calls to servlet API calls such as getCallerPrincipal, for example, will still return null even though there is actually an anonymous authentication object in the SecurityContextHolder.
* Symfony uses AnonymousToken much more than "just for convience in access-control attributes". **Removing anonymous tokens allows us to move towards only allowing `UserInterface` users**: #34909

---

Removing anonymous tokens do have an impact on `AccessListener` and `AuthorizationChecker`. These currently throw an exception if there is no token in the storage, instead of treating them like "unknown users" (i.e. no roles). See #30609 on a RFC about removing this exception. We can also see e.g. the [Twig `is_granted()` function explicitly catching this exception](https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/Extension/SecurityExtension.php#L37-L52).

* **To make the changes in `AccessListener` and `AuthorizationChecker` BC, a flag has been added - default enabled - to throw an exception when no token is present** (which is automatically disabled when the new system is used). In Symfony 5.4 (or whenever the new system is no longer experimental), we can deprecate this flag and in 6.0 we can never throw the exception anymore.
* **`anonymous: lazy` has been deprecated in favor of `{ anonymous: true, lazy: true }`** This fixes the dependency on `AnonymousFactory` from the `SecurityExtension` and allows removing the `anonymous` option.
* **Introduced `PUBLIC_ACCESS` Security attribute** as alternative of `IS_AUTHENTICATED_ANONYMOUSLY`. Both work in the new system, the latter only triggers a deprecation notice (but may be usefull to allow switching back and forth between old and new system).

cc @javiereguiluz you might be interested, as I recently talked with you about this topic

### CODE REVIEW

**Comment by simonberger** [comment]
I really like this change. The anonymous part of the authentication process was always troublesome to handle. Your changes in #30914 are generally much appreciated.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php [hunk 1]
```diff
@@ -72,12 +74,20 @@ public function authenticate(RequestEvent $event)
         $attributes = $request->attributes->get('_access_control_attributes');
         $request->attributes->remove('_access_control_attributes');
 
-        if (!$attributes || ([AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY] === $attributes && $event instanceof LazyResponseEvent)) {
+        if (!$attributes || [AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY] === $attributes) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php:77]: I don't understand the internals of this ... but are we going to keep this `IS_AUTHENTICATED_ANONYMOUSLY`permission? I hope we can get rid of everything about "anonymous + authenticated".

**Comment by javiereguiluz** [comment]
@wouterj thanks for this 🙇  Trying to explain that "anonymous users" are "authenticated" was one of the most WTF moments in Symfony training. It never made sense to anyone. So glad to get rid of this.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php (hunk 1 above)
Comment (weaverryan) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php:77]: There DOES still need to be something like this, as it’s used when you are “whitelisting” paths in access_control. But it doesn’t need to have this name... though changing the name could be done later - that may simplify things.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/AnonymousFactory.php [hunk 2]
```diff
@@ -44,16 +45,7 @@ public function create(ContainerBuilder $container, $id, $config, $userProvider,
...
-            $config['secret'] = new Parameter('container.build_hash');
-        }
-
-        $authenticatorId = 'security.authenticator.anonymous.'.$firewallName;
-        $container
-            ->setDefinition($authenticatorId, new ChildDefinition('security.authenticator.anonymous'))
-            ->replaceArgument(0, $config['secret']);
-
-        return $authenticatorId;
+        throw new InvalidConfigurationException('The authenticator manager no longer has anonymous security.');
```
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/AnonymousFactory.php:48]: Quotes around “anonymous” and maybe the call to action:

Please remove this option under the “main” firewall.

Probably once we get the lazy firewall thing working, this might need to say something like “remove the anonymous option under the main firewall and add lazy: true instead”.

File: src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php [hunk 3]
```diff
@@ -57,23 +57,20 @@ public function supports(Request $request): ?bool
...
         return null;
     }
 
     public function authenticate(Request $request): PassportInterface
     {
-        $token = $this->rememberMeServices->autoLogin($request);
-
-        return new SelfValidatingPassport($token->getUser());
+        return new SelfValidatingPassport($request->attributes->get('_remember_me_token')->getUser());
     }
```
Comment (weaverryan) [inline src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php:74]: Are these changes related? If so, what’s causing them?

Comment (wouterj) [reply src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php:74]: Yes, it is. `$token` was `null`, causing an error here. I don't remember anymore why exactly (and am not on my dev machine for the coming days).

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php (hunk 1 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php:77]: Yeah, so we need an attribute to whitelist pages that are within the firewall, but accessible for unauthenticated users (e.g. the login page). 

A couple suggestions after a chat about this:
1. `PUBLIC_ACCESS`
3. `ALLOW_ACCESS_TO_ALL`
2. `NO_SECURITY`

Do you, @javiereguiluz, or someone else favor one of these options (or another suggestion)?

Comment (simonberger) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php:77]: Also that we want to get rid of anonymous authentication the trigger word anonymous is still very good to describe an unprotected access. So another option could be ANONYMOUS_ACCESS in this context.

Comment (javiereguiluz) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php:77]: Random idea ... add a `exclude` option for `access_control` entries:

```yaml
access_control:
    - { path: '^/admin/users', roles: ROLE_SUPER_ADMIN }
    - { path: '^/admin', roles: ROLE_ADMIN, exclude: ['^/admin/login', '^/admin/public/*'] }
```

File: src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php (hunk 3 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Authenticator/RememberMeAuthenticator.php]: Reverted this and no longer had problems. I still think these changes make sense (according to the signature, `$token` can still be null here), but let's not do that in this PR :)

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php (hunk 1 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: Thanks for thinking outside the box @javiereguiluz! I'm not totally sure if this would make things much clearer, given the access control is already a list of patterns. Having a sub-list of excluded pattern for each pattern might make it more confusing to see which rule is used for which request. Maybe we have to completely revisit the access control configuration in another version of Symfony :smile: 

I've updated the PR to use `PUBLIC_ACCESS`. Using anonymous is correct, it's maybe best to (at least for the next versions) avoid that term to remove confusion. Also, any logged in user is not accessing such a page "anonymously".

**Comment by wouterj** [comment]
This PR is now tested in the Symfony Demo application and works like charm :tada: . I've updated the PR description with the current state. Ready for review & merge imho.

---

I managed to break fabbot's exception message format and I'm unable to fix the error using concatenation :cry: The other build failures are also unrelated.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @wouterj.

**Comment by kappa1389** [comment]
If token is not provided in request and no access control is defined for called route then firewall gets bypassed, but the expected  behavior is to get an authentication exception with code of 401
I face this problem with new symfony security integrated with lexik
Any ideas?

**Comment by derrabus** [comment]
Hello @kappa1389 and thank you for reaching out.

This tracker is meant for tracking bugs and feature requests. If you are looking for support, you'll get faster responses by using one of our official support channels:

* StackOverflow: https://stackoverflow.com/questions/tagged/symfony
* Slack: https://symfony.com/slack

See https://symfony.com/support for more support options.

**Comment by kappa1389** [comment]
Dear Alexander
Thank you for your reply i will take it into account

Best regards
